### PR TITLE
support fields from multiple remote schemas in the same query (fix #1371)

### LINKED ISSUE #1371: support mixing top-level fields from two different schemas
After a remote schema is added, currently Hasura does not allow mixing root fields from different schemas. Add support for it.

E.g:

```graphql
query {
  profile {
    id
    name
  }
  weather(city: "BLR") {
    temperature
  }
}
```

Here `profile` is from Hasura schema (a table). And `weather` is from a remote schema.

### PR DESCRIPTION
**Pull request by ecthiender**
Support fields from multiple graphql servers in the same query.

### Description

Support multiple fields from different graphql servers in a single query. This removes existing limitation.

**Example:**

```graphql
query ($title: String!, $code: String!) {
  author  {
    name
    articles (where: {title: {_like: $title}}) {
      title
      content
    }
  }

  country(code: $code) {
    name
  }
  langs: programming_languages {
    lang: name
  }
}
```
Where `author` and `article` are from hasura, `country` and `programming_langauges` are from a remote schema.

Response:
```json
{
  "data": [
    {
      "author": [
        {
          "name": "jane",
          "articles": [
            {
              "content": "content for the first post",
              "title": "first post"
            }
          ]
        },
        {
          "name": "john",
          "articles": [
            {
              "content": "lorem ipsum lorem ipsum",
              "title": "second title"
            }
          ]
        },
        {
          "name": "asd",
          "articles": []
        },
        {
          "name": "new one",
          "articles": []
        },
        {
          "name": "new one 1",
          "articles": []
        }
      ]
    },
    {
      "country": {
        "name": "India"
      }
    },
    {
      "langs": [
        {
          "lang": "haskell"
        },
        {
          "lang": "rust"
        }
      ]
    }
  ],
  "errors": []
}
```

### Affected components 

- Server

### Related Issues

#1371

### Solution and Design

TODO:
1. errors should indicate from which remote it comes
2. support ws transport

### Steps to test and verify

On the console:
1. Create two tables  `author` and `article`. Create relationships, insert sample data.
2. Add the following remote schema: https://countries.trevorblades.com/
3. Make the following query:

```graphql
query ($title: String!, $code: String!) {
  author  {
    name
    articles {
      title
      content
    }
  }

  country(code: $code) {
    name
  }
```

### Limitations, known bugs & workarounds

This [PR](https://github.com/hasura/graphql-parser-hs/pull/12) has to be merged before merging this.

### CODE REVIEW

File: server/src-lib/Hasura/GraphQL/Transport/HTTP.hs [hunk 1]
```diff
@@ -26,15 +28,49 @@ runGQ
...
+
+  let interimResBS = map encJToLBS resSet
+  interimRes <- forM interimResBS $ \res -> do
+    let x = J.decode res :: (Maybe J.Object)
+    onNothing x $ throw500 "could not parse response as JSON"
+
+  let datas = mapMaybe (Map.lookup "data") interimRes
+      errs  = mapMaybe (Map.lookup "errors") interimRes
+
+  return $ encJFromJValue $ J.object [ "data" J..= datas
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Transport/HTTP.hs]: This returns an array of data which is not correct, it should return an object with key value pairs.

Comment (ecthiender) [reply server/src-lib/Hasura/GraphQL/Transport/HTTP.hs]: Fixed it.

File: server/src-lib/Hasura/GraphQL/Transport/HTTP.hs [hunk 2]
```diff
@@ -26,15 +28,56 @@ runGQ
...
+                                     , "errors" J..= errs
+                                     ]
+
+  where
+    -- TODO: should validate response and throw error?
+    onlyObjs jVals =
+      let fn jVal = case jVal of
+            J.Object o -> Just o
+            _          -> Nothing
+      in mapMaybe fn jVals
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Transport/HTTP.hs]: Why not use filter?

File: server/src-lib/Hasura/GraphQL/Transport/HTTP.hs [hunk 3]
```diff
@@ -26,15 +28,56 @@ runGQ
...
+    E.GExPMixed _ ->
+      throw500 "internal-unexpected: mixed plan is nested in mixed plan"
+
+  let interimResBS = map encJToLBS resSet
+  interimRes <- forM interimResBS $ \res -> do
+    let x = J.decode res :: (Maybe J.Object)
+    onNothing x $ throw500 "could not parse response as JSON"
+
+  let datas = onlyObjs $ mapMaybe (Map.lookup "data") interimRes
+      errs  = mapMaybe (Map.lookup "errors") interimRes
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Transport/HTTP.hs]: errs should be Maybe type so that you don't get the `errors` key if the errors array is empty

**Comment by 0x777** [comment]
Closing this in favour of #3153

**Pull request closed without merging**